### PR TITLE
fix effect_free computation over getfield

### LINKED ISSUE #18015: Inlining(?) bug on 0.5
The following code works on julia 0.4 but fails on 0.5-rc2. It does, however, work when run through Gallium or when `setabc2!` is annotated with `@noinline`.

```jl
using Base.Test
import Base: ==

type Triple
  a::Int
  b::Int
  c::Int
end

a(tri) = tri.a
b(tri) = tri.b
c(tri) = tri.c
setabc!(tri, a, b, c) = (tri.a = a; tri.b = b; tri.c = c)
setabc2!(tri, a, b, c) = (tri.a = a; tri.b = b; tri.c = c; nothing)
==(tri1::Triple, tri2::Triple) = ((tri1.a, tri1.b, tri1.c) == (tri2.a, tri2.b, tri2.c))

function main()
  tri = Triple(1, 2, 3)
  tri2 = Triple(1, 2, 3)
  setabc!(tri, b(tri), c(tri), a(tri))
  setabc2!(tri2, b(tri2), c(tri2), a(tri2))
  @test tri == tri2
end

main()


```

### PR DESCRIPTION
**Pull request by vtjnash**
fix #18015 

@JeffBezanson this also fixes the bug you noted where inlineable was passing the wrong `sv` to `effect_free`. It would be appreciated if you could verify I've picked the right linfo to pass in each case.

### CODE REVIEW

**Comment by tkelman** [comment]
osx travis got stuck, log backed up to https://gist.github.com/a62ec9f360ea4b4c8a8cdaec465b52dd and restarted

File: base/inference.jl [hunk 1]
```diff
@@ -2251,32 +2252,17 @@ function effect_free(e::ANY, sv, allow_volatile::Bool)
...
-                                end
-                            end
-                            if !effect_free(a,sv,allow_volatile)
-                                return false
-                            end
+                    elseif is_known_call(e, getfield, linfo)
+                        # first argument must be immutable to ensure e is affect_free
+                        a = ea[2]
+                        typ = widenconst(exprtype(a, linfo))
+                        if !isa(typ, DataType) || typ.mutable
```
Comment (JeffBezanson) [inline base/inference.jl]: I think we might need `|| typ.abstract` here.

File: base/inference.jl [hunk 2]
```diff
@@ -2614,16 +2600,16 @@ function inlineable(f::ANY, ft::ANY, e::Expr, atypes::Vector{Any}, sv::Inference
                 occ += occurs_more(b, x->(isa(x,Slot)&&x.id==i), 6)
             end
             # TODO: passing `sv` here is wrong since it refers to the enclosing function
```
Comment (JeffBezanson) [inline base/inference.jl]: Comment can be deleted.

**Comment by JeffBezanson** [comment]
> if you could verify I've picked the right linfo to pass in each case

Yes that part looks right to me.

**Comment by tkelman** [comment]
It looks like this caused some pretty major performance regressions.

https://github.com/JuliaCI/BaseBenchmarkReports/blob/1ea0d8c99a047f5d72c9d2aa70af49ccc8ab5463/c48d203_vs_b80dad4/report.md is mostly fine at c48d203702f7b439792ffd9d24abd23e47de171b,

https://github.com/JuliaCI/BaseBenchmarkReports/blob/ec71a5c4cb488dde420d1908f81b7b66524630c8/a0d75a4_vs_b80dad4/report.md is bad at a0d75a49ceb6ea31b64fc85707916dc0b3e9300e

**Comment by KristofferC** [comment]
Can we try to use this super awesome benchmark tool a bit more frequently. A run doesn't take that long. Saves @tkelman  from manually having to bisect performance regressions.

**Comment by JeffBezanson** [comment]
I agree. Running benchmarks on every commit is starting to get urgent. If it takes too long or we don't have enough resources, we should run a reduced version of the suite. Running on every commit is much more important than having perfect coverage.

**Pull request merged**